third_party/move/tools/move-unit-test/src/test_runner.rs

```diff
@@ -255,71 +234,60 @@ impl<'a, 'b, W: Write> TestOutput<'a, 'b, W> {
-        VMResult<Vec<Vec<u8>>>,
-        TestRunInfo,
-    ) {
+        factory: &Mutex<F>,
+    ) -> (VMResult<()>, VMResult<ChangeSet>, TestRunInfo) {
         let move_vm = MoveVM::new(self.native_function_table.clone());
         let extensions = extensions::new_extensions();
         let mut session =
             move_vm.new_session_with_extensions(&self.starting_storage_state, extensions);
-        let mut gas_meter = GasStatus::new(&self.cost_table, Gas::new(self.execution_bound));
+        let mut gas_meter = factory.lock().unwrap().new_gas_meter();
+
         // TODO: collect VM logs if the verbose flag (i.e, `self.verbose`) is set
 
         let now = Instant::now();
         let storage = TraversalStorage::new();
-        let serialized_return_values_result = session.execute_function_bypass_visibility(
-            &test_plan.module_id,
-            IdentStr::new(function_name).unwrap(),
-            vec![], // no ty args, at least for now
-            serialize_values(test_info.arguments.iter()),
-            &mut gas_meter,
-            &mut TraversalContext::new(&storage),
-        );
-        let mut return_result = serialized_return_values_result.map(|res| {
-            res.return_values
-                .into_iter()
-                .map(|(bytes, _layout)| bytes)
-                .collect()
-        });
+        let mut execute_result = session
+            .execute_function_bypass_visibility(
+                &test_plan.module_id,
+                IdentStr::new(function_name).unwrap(),
+                vec![], // no ty args, at least for now
+                serialize_values(test_info.arguments.iter()),
+                &mut gas_meter,
+                &mut TraversalContext::new(&storage),
+            )
+            .map(|_| ());
         if !self.report_stacktrace_on_abort {
-            if let Err(err) = &mut return_result {
+            if let Err(err) = &mut execute_result {
                 err.remove_exec_state();
             }
         }
-        let test_run_info = TestRunInfo::new(
-            function_name.to_string(),
-            now.elapsed(),
-            // TODO(Gas): This doesn't look quite right...
-            //            We're not computing the number of instructions executed even with a unit gas schedule.
-            Gas::new(self.execution_bound)
-                .checked_sub(gas_meter.remaining_gas())
-                .unwrap()
-                .into(),
+
+        let test_run_info = TestRunInfo::new(function_name.to_string(), now.elapsed());
+        let (cs_result, test_run_info) = factory.lock().unwrap().finish_session(
+            session,
+            gas_meter,
+            test_run_info,
         );
-        match session.finish_with_extensions() {
-            Ok((cs, extensions)) => (Ok(cs), Ok(extensions), return_result, test_run_info),
-            Err(err) => (Err(err.clone()), Err(err), return_result, test_run_info),
-        }
+
+        (execute_result, cs_result, test_run_info)
```

Comment:
Yeah, as I mentioned, let's keep the old finish session logic
```
        match session.finish_with_extensions() {
            Ok((cs, extensions)) => (Ok(cs), Ok(extensions), return_result, test_run_info),
Comment
            Err(err) => (Err(err.clone()), Err(err), return_result, test_run_info),
        }
```
but add a step after to finalize the test info.